ingestion/src/main.rs

```diff
@@ -66,25 +67,41 @@ async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
         }
     });
 
-    // Spawn kvkafka reader
-    #[cfg(feature = "kafka_prod")]
-    {
-        let kafka_group = args[1].to_string();
-        println!("Spawing kvkafka reader...");
-        tokio::spawn(lard_ingestion::kvkafka::read_and_insert(
-            db_pool.clone(),
-            kafka_group,
-        ));
-    }
+    // set up cancellation token and signal catcher for graceful shutdown
+    let cancel_token = CancellationToken::new();
+    tokio::spawn(util::signal_catcher(cancel_token.clone()));
 
     // Set up and run our server + database
     println!("Ingestion server started!");
-    lard_ingestion::run(
+    let ingestor = tokio::spawn(lard_ingestion::run(
         db_pool,
         PARAMCONV,
         permit_tables,
         rove_connector,
         qc_pipelines,
-    )
-    .await
+        cancel_token,
+    ));
+
+    #[cfg(feature = "kafka_prod")]
+    // Spawn kvkafka reader
+    {
+        let kafka_group = args[1].to_string();
+        println!("Spawning kvkafka reader...");
+        let kvkafka_reader = tokio::spawn(lard_ingestion::kvkafka::read_and_insert(
+            db_pool.clone(),
+            kafka_group,
+            cancel_token.clone(),
+        ));
+
+        let (ingestor_res, kvkafka_reader_res) = tokio::join!(ingestor, kvkafka_reader);
+        (_, _) = (ingestor_res, kvkafka_reader_res); // ignore for now
+    }
+
+    #[cfg(not(feature = "kafka_prod"))]
+    {
+        let ingestor_res = tokio::join!(ingestor);
+        _ = ingestor_res; // ignore for now
+    }
+
+    Ok(())
```

Comment:
This has changed from returning a result from the main function to ignoring it. Not a big deal, but since the results are there I think we should use them.

In the `kafka_prod` case you can combine the two results with `.and_then()`

Reply:
I'll take a look (after lunch).

Reply:
Done!